### PR TITLE
Solution Crash Ipls and Improvements

### PR DESCRIPTION
**Pull request by YimMenuTheBest**
Solution to ipls crash and Improved implementation of ipls unloading and loading when switching ipls

### CODE REVIEW

**Comment by YimMenuTheBest** [comment]
This was inspired by Tessio and Issue #1505

File: src/views/self/view_teleport.cpp [hunk 1]
```diff
@@ -110,20 +110,40 @@ namespace big
...
+				g_fiber_pool->queue_job([&] {
+					STREAMING::REMOVE_IPL(ipl_name_remove_when_load);
+					LOG(INFO) << "ipls removed to avoid conflicts " << ipl_name_remove_when_load;
+				});
+
 			//unload all previous ipls
 			for (auto& ipl : ipls)
 				for (auto& ipl_name : ipl.ipl_names)
 				{
 					if (STREAMING::IS_IPL_ACTIVE(ipl_name))
```
Comment (gir489returns) [inline src/views/self/view_teleport.cpp:124]: Any native must be called within a fiber, you can't call it from the presentation layer. This was probably there before you touched the code, but it's still wrong and needs to be fixed. The entire code block should be moved into the job, as you're basically queuing up X number of fibers for each active IPL that needs to be removed. One action -> one fiber.

File: src/views/self/view_teleport.cpp [hunk 2]
```diff
@@ -110,20 +110,40 @@ namespace big
...
+								LOG(INFO) << "loaded all previously deleted ipls " << ipl_name_remove;
+							}
+						});
 					}
+					
 				}
 
-			//load the new ipl
+			//load the new ipls
 			for (auto& ipl_name : selected_ipl.ipl_names)
```
Comment (gir489returns) [inline src/views/self/view_teleport.cpp:140]: Ditto.

**Review by gir489returns — CHANGES_REQUESTED** [review]

**Comment by YimMenuTheBest** [comment]
At this point I just realized that all the ipls are loading and unloading when loading a single one ipl, in conclusion... skill issues

**Comment by Tessio** [comment]
> At this point I just realized that all the ipls are loading and unloading when loading a single one ipl, in conclusion... skill issues

*Definitive
```
static int current_select = g.self.ipls.select;
		static int last_select    = current_select;

		if (ImGui::BeginCombo("IPL_LOCATION"_T.data(), ipls[current_select].friendly_name))
		{
			for (int i = 0; i < IM_ARRAYSIZE(ipls); i++)
			{
				bool is_selected = (i == current_select);
				if (ImGui::Selectable(ipls[i].friendly_name, is_selected))
				{
					current_select = i;
				}
				if (is_selected)
				{
					ImGui::SetItemDefaultFocus();
				}
			}
			ImGui::EndCombo();
		}

		if (ImGui::Button("LOAD_IPL"_T.data()))
		{
			g_fiber_pool->queue_job([&] {
				// If we've changed selections, first unload previously loaded IPL, then load previously deleted IPLs
				if (current_select != last_select)
				{
					// Unload previously loaded IPL of the last selection
					for (auto& ipl_name_unload : ipls[last_select].ipl_names)
					{
						if (STREAMING::IS_IPL_ACTIVE(ipl_name_unload))
						{
							STREAMING::REMOVE_IPL(ipl_name_unload);
							LOG(INFO) << "Unloaded previous IPL " << ipl_name_unload;
						}
					}

					// Delay to wait for the unloading to complete
					script::get_current()->yield(100ms);

					// Load previously deleted IPLs of the last selection
					for (auto& ipl_name_load : ipls[last_select].ipl_names_remove)
					{
						STREAMING::REQUEST_IPL(ipl_name_load);
						LOG(INFO) << "Loaded previously deleted IPLs " << ipl_name_load;
					}

					// Small delay to allow for IPLs to load
					script::get_current()->yield(100ms);

					// Load new IPLs of the current selection
					for (auto& ipl_name : ipls[current_select].ipl_names)
					{
						STREAMING::REQUEST_IPL(ipl_name);
						LOG(INFO) << "New IPL loaded " << ipl_name;
					}

					// Remove old IPLs of the current selection to avoid conflicts
					for (auto& ipl_name_remove : ipls[current_select].ipl_names_remove_when_load)
					{
						STREAMING::REMOVE_IPL(ipl_name_remove);
						LOG(INFO) << "Removed IPL to avoid conflict " << ipl_name_remove;
					}

					last_select = current_select;
				}
			});
		}

		ImGui::SameLine();

		if (components::button("TP_TO_IPL"_T.data()))
		{
			g_fiber_pool->queue_job([&] {
				teleport::to_coords(ipls[current_select].location); // Takes the location of the selected IPL
			});
		}

		ImGui::Spacing();
		components::small_text("IPL_INFOS"_T.data());

		ImGui::Text(std::vformat("IPL_CNT"_T, std::make_format_args(ipls[current_select].ipl_names.size())).data());
		ImGui::Text(std::vformat("IPL_POSITION"_T,
		    std::make_format_args(ipls[current_select].location.x,
		        ipls[current_select].location.y,
		        ipls[current_select].location.z))
		                .data());
	}
}
```

File: src/views/self/view_teleport.cpp [hunk 3]
```diff
@@ -109,28 +109,48 @@ namespace big
 
 		const auto& selected_ipl = ipls[g.self.ipls.select];
 		if (components::button("LOAD_IPL"_T.data()))
-		{
-			//unload all previous ipls
-			for (auto& ipl : ipls)
-				for (auto& ipl_name : ipl.ipl_names)
-				{
-					if (STREAMING::IS_IPL_ACTIVE(ipl_name))
+		{
```
Comment (gir489returns) [inline src/views/self/view_teleport.cpp:112]: components::button has a method of invoking queuing a fiber without explicitly calling queue_job. See this example here:  https://github.com/YimMenu/YimMenu/blob/d6c32533be18aef5b35122b4f068b8844155751c/src/views/self/view_weapons.cpp#L249-L251

**Review by gir489returns — CHANGES_REQUESTED** [review]

File: src/views/self/view_teleport.cpp (hunk 3 above)
Comment (gir489returns) [reply src/views/self/view_teleport.cpp:112]: Also, components::button takes a string_view as a parameter, there is no need to invoke the .data() accessor on it to convert it to a const char*, which will then be re-cast to a string_view. You're basically doing string_view (translation) -> const char* (.data()) -> string_view (parameter) -> const char* (components::button's invoke of ImGui)

**Comment by YimMenuTheBest** [comment]
> > At this point I just realized that all the ipls are loading and unloading when loading a single one ipl, in conclusion... skill issues
> 
> It's not perfect, but it could work
> 
> ```
>static int current_select = g.self.ipls.select;
		static int last_select    = current_select;

		if (ImGui::BeginCombo("IPL_LOCATION"_T.data(), ipls[current_select].friendly_name))
		{
			for (int i = 0; i < IM_ARRAYSIZE(ipls); i++)
			{
				bool is_selected = (i == current_select);
				if (ImGui::Selectable(ipls[i].friendly_name, is_selected))
				{
					current_select = i;
				}
				if (is_selected)
				{
					ImGui::SetItemDefaultFocus();
				}
			}
			ImGui::EndCombo();
		}

		if (ImGui::Button("LOAD_IPL"_T.data()))
		{
			g_fiber_pool->queue_job([&] {
				// If we've changed selections, first unload previously loaded IPL, then load previously deleted IPLs
				if (current_select != last_select)
				{
					// Unload previously loaded IPL of the last selection
					for (auto& ipl_name_unload : ipls[last_select].ipl_names)
					{
						if (STREAMING::IS_IPL_ACTIVE(ipl_name_unload))
						{
							STREAMING::REMOVE_IPL(ipl_name_unload);
							LOG(INFO) << "Unloaded previous IPL " << ipl_name_unload;
						}
					}

					// Delay to wait for the unloading to complete
					script::get_current()->yield(100ms);

					// Load previously deleted IPLs of the last selection
					for (auto& ipl_name_load : ipls[last_select].ipl_names_remove)
					{
						STREAMING::REQUEST_IPL(ipl_name_load);
						LOG(INFO) << "Loaded previously deleted IPLs " << ipl_name_load;
					}

					// Small delay to allow for IPLs to load
					script::get_current()->yield(100ms);

					// Load new IPLs of the current selection
					for (auto& ipl_name : ipls[current_select].ipl_names)
					{
						STREAMING::REQUEST_IPL(ipl_name);
						LOG(INFO) << "New IPL loaded " << ipl_name;
					}

					// Remove old IPLs of the current selection to avoid conflicts
					for (auto& ipl_name_remove : ipls[current_select].ipl_names_remove_when_load)
					{
						STREAMING::REMOVE_IPL(ipl_name_remove);
						LOG(INFO) << "Removed IPL to avoid conflict " << ipl_name_remove;
					}

					last_select = current_select;
				}
			});
		}

		ImGui::SameLine();

		if (components::button("TP_TO_IPL"_T.data()))
		{
			g_fiber_pool->queue_job([&] {
				teleport::to_coords(ipls[current_select].location); // Toma la ubicacin de la IPL seleccionada
			});
		}

		ImGui::Spacing();
		components::small_text("IPL_INFOS"_T.data());

		ImGui::Text(std::vformat("IPL_CNT"_T, std::make_format_args(ipls[current_select].ipl_names.size())).data());
		ImGui::Text(std::vformat("IPL_POSITION"_T,
		    std::make_format_args(ipls[current_select].location.x,
		        ipls[current_select].location.y,
		        ipls[current_select].location.z))
		                .data());
	}
}
> ```

It meets the main idea of pull, thanks @Tessio for complementing and to @gir489returns for being a great person ♥

- No more crashes for not being in Fiber Pool.

- Does not generate invoker error when teleporting

- Restore ipls

- Removes ipls when switching to another

- And it does not reload all ipls when selecting another one.

**Review by Tessio — COMMENTED** [review]
Works smooth as butter uwu/
image

File: src/views/self/view_teleport.cpp [hunk 4]
```diff
@@ -93,52 +93,82 @@ namespace big
...
+		components::button("TP_TO_IPL"_T.data(), [] 
 		{
-			teleport::to_coords(selected_ipl.location);
-		}
+			teleport::to_coords(ipls[current_select].location);
+		});
 
 		ImGui::Spacing();
-		components::small_text("IPL_INFOS"_T);
+		components::small_text("IPL_INFOS"_T.data());
```
Comment (gir489returns) [inline src/views/self/view_teleport.cpp:169]: Remove extraneous .data() call.

File: src/views/self/view_teleport.cpp [hunk 5]
```diff
@@ -93,52 +93,82 @@ namespace big
...
+				{
+					STREAMING::REMOVE_IPL(ipl_name_remove);
+					LOG(INFO) << "Removed IPL to avoid conflict " << ipl_name_remove;
+				}
+
+				last_select = current_select;
+			}
+		});
+		ImGui::SameLine();
+		components::button("TP_TO_IPL"_T.data(), []
```
Comment (gir489returns) [inline src/views/self/view_teleport.cpp:163]: Remove extraneous .data() call.

File: src/views/self/view_teleport.cpp [hunk 6]
```diff
@@ -93,52 +93,82 @@ namespace big
...
+				}
 			}
-
 			ImGui::EndCombo();
 		}
-
-		const auto& selected_ipl = ipls[g.self.ipls.select];
-		if (components::button("LOAD_IPL"_T))
+		ImGui::SameLine();
+		components::button("LOAD_IPL"_T.data(), []
```
Comment (gir489returns) [inline src/views/self/view_teleport.cpp:117]: Remove extraneous .data() call.

File: src/core/data/ipls.hpp [hunk 7]
```diff
@@ -5,18 +5,79 @@ struct ipl_libary
...
+        {{"post_hiest_unload"}},
+        {{"jewel2fake"}},
+        {{"jewel2fake"}, {"bh1_16_refurb"}},
+    }, // Jewelry Interior
+    {"Coroner Morgue Interior", 
+    {244.9, -1374.7, 39.5},
+    {{"coroner_int_on"}, {"coronertrash"}},
+    {{"coroner_int_off"}},
+    {{"coroner_int_off"}},
+    }, // Coroner Morgue Interior
```
Comment (gir489returns) [inline src/core/data/ipls.hpp:82]: There's a lot more IPLs you could load, would be actually useful like the Arcade or Casino, but that's an issue for a different PR.

**Review by gir489returns — CHANGES_REQUESTED** [review]

File: src/core/data/ipls.hpp (hunk 7 above)
Comment (YimMenuTheBest) [reply src/core/data/ipls.hpp:82]: I will add all the ipls, in a later pr c:

File: src/views/self/view_teleport.cpp [hunk 8]
```diff
@@ -93,52 +93,82 @@ namespace big
...
-		ImGui::SameLine();
+				// Load previously deleted IPLs of the last selection
+				for (auto& ipl_name_load : ipls[last_select].ipl_names_remove)
+				{
+					STREAMING::REQUEST_IPL(ipl_name_load);
+					LOG(INFO) << "Loaded previously deleted IPLs " << ipl_name_load;
+				}
+
+				// Small delay to allow for IPLs to load
+				script::get_current()->yield(100ms);
```
Comment (gir489returns) [inline src/views/self/view_teleport.cpp:143]: Is this required? I don't see any other script do this when handling IPLs, they don't seem to care if it got removed or not, should we? I think simply calling REMOVE_IPL is enough, unless it's causing collisions with other IPLs.

Comment (YimMenuTheBest) [reply src/views/self/view_teleport.cpp]: My bad, this is no longer necessary

**Review by gir489returns — APPROVED** [review]
You earned this.

File: src/views/self/view_teleport.cpp [hunk 9]
```diff
@@ -93,52 +93,76 @@ namespace big
...
+				{
+					STREAMING::REQUEST_IPL(ipl_name);
+					LOG(INFO) << "New IPL loaded " << ipl_name;
+				}
+
+				// Remove old IPLs of the current selection to avoid conflicts
+				for (auto& ipl_name_remove : ipls[current_select].ipl_names_remove_when_load)
+				{
+					STREAMING::REMOVE_IPL(ipl_name_remove);
+					LOG(INFO) << "Removed IPL to avoid conflict " << ipl_name_remove;
```
Comment (Yimura) [inline src/views/self/view_teleport.cpp]: These logs are likely to cause spam, best to remove these before adding it to master.

**Review by Yimura — CHANGES_REQUESTED** [review]
Looks good, only one requested change.

File: src/views/self/view_teleport.cpp (hunk 9 above)
Comment (YimMenuTheBest) [reply src/views/self/view_teleport.cpp]: solved

**Review by Yimura — APPROVED** [review]
Thanks for contributing to YimMenu!

**Comment by gir489returns** [comment]
> Thanks for contributing to YimMenu!

Hopefully the first of many.

**Comment by YimMenuTheBest** [comment]
A pleasure to contribute, may there be many more!

**Pull request merged**